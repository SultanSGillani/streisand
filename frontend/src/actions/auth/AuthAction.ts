import { replace } from 'react-router-redux';

import Store from '../../store';
import globals from '../../utilities/globals';
import { post } from '../../utilities/Requestor';
import { IUnkownError } from '../../models/base/IError';
import { ThunkAction, IDispatch } from '../ActionTypes';
import { storeAuthToken } from '../../utilities/storage';
import ErrorAction, { handleError } from '../ErrorAction';

type AuthAction =
    { type: 'AUTHENTICATING' } |
    { type: 'AUTHENTICATED', token: string } |
    { type: 'AUTHENTICATION_FAILED', message: string };
export default AuthAction;
type Action = AuthAction | ErrorAction;

function authenticating(): Action {
    return { type: 'AUTHENTICATING' };
}

function authenticated(token: string): Action {
    return {
        type: 'AUTHENTICATED',
        token: token
    };
}

export function login(username: string, password: string): ThunkAction<Action> {
    return (dispatch: IDispatch<Action>, getState: () => Store.All) => {
        const state = getState();
        dispatch(authenticating());
        return request(username, password).then((result: { token: string }) => {
            storeAuthToken(result.token);
            const action = dispatch(authenticated(result.token));
            if (state.location.referrer) {
                dispatch(replace(state.location.referrer));
            } else {
                dispatch(replace('/'));
            }
            return action;
        }, (error: IUnkownError) => {
            return dispatch(handleError(error, 'Login failed'));
        });
    };
}

function request(username: string, password: string): Promise<{ token: string }> {
    const data = { username, password };
    return post({ data, url: `${globals.apiUrl}/login/` });
}
