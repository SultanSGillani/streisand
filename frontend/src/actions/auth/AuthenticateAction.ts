import { replace } from 'react-router-redux';
import { put, select } from 'redux-saga/effects';

import Store from '../../store';
import globals from '../../utilities/globals';
import { post } from '../../utilities/Requestor';
import { storeAuthToken } from '../../utilities/storage';
import { generateFetch, generateSage } from '../sagas/generators';

interface IActionProps {
    username: string;
    password: string;
}

export type RequestAuthentication = { type: 'REQUEST_AUTHENTICATION', props: IActionProps };
export type ReceivedAuthentication = { type: 'RECEIVED_AUTHENTICATION', token: string };
export type FailedAuthentication = { type: 'FAILED_AUTHENTICATION' };

type AuthenticateAction = RequestAuthentication | ReceivedAuthentication | FailedAuthentication;
export default AuthenticateAction;
type Action = AuthenticateAction;

type AuthResponse = { token: string; };

function* received(response: AuthResponse) {
    storeAuthToken(response.token);
    yield put<Action>({ type: 'RECEIVED_AUTHENTICATION', token: response.token });
    const state: Store.All = yield select();
    if (state.location.referrer) {
        yield put(replace(state.location.referrer));
    } else {
        yield put(replace('/'));
    }
}

function failure(): Action {
    return { type: 'FAILED_AUTHENTICATION' };
}

export function login(username: string, password: string): Action {
    return { type: 'REQUEST_AUTHENTICATION', props: { username, password } };
}

const errorPrefix = 'Authentication failed';
const fetch = generateFetch({ errorPrefix, request, received, failure });
export const authenticateSaga = generateSage<RequestAuthentication>('REQUEST_AUTHENTICATION', fetch);

function request(data: IActionProps): Promise<AuthResponse> {
    return post({ data, url: `${globals.apiUrl}/login/` });
}
