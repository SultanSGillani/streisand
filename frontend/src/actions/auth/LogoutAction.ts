
import { storeAuthToken } from '../../utilities/storage';

type LogoutAction = { type: 'LOGOUT' };
export default LogoutAction;
type Action = LogoutAction;

export function logout(): Action {
    storeAuthToken('');
    return { type: 'LOGOUT' };
}