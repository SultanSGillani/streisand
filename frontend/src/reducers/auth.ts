import * as assign from 'object-assign';

import Action from '../actions/auth';
import IAuthInfo from '../models/IAuthInfo';
import { getAuthToken } from '../utilities/storage';

const authToken = getAuthToken();
const defaultValue: IAuthInfo = {
    isAuthenticated: !!authToken,
    isAuthenticating: false,
    token: authToken
};

function auth(state: IAuthInfo = defaultValue, action: Action): IAuthInfo {
    switch (action.type) {
        case 'LOGOUT':
        case 'AUTHENTICATION_FAILED':
            return {
                isAuthenticated: false,
                isAuthenticating: false,
                token: ''
            };
        case 'AUTHENTICATING':
            return assign({}, state, { isAuthenticating: true });
        case 'AUTHENTICATED':
            return {
                isAuthenticated: true,
                isAuthenticating: false,
                token: action.token
            };
        default:
            return state;
    }
}

export default auth;