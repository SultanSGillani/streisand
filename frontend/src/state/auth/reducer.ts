import * as assign from 'object-assign';

import Action from './actions';
import { IAuthStore } from './store';
import { getAuthToken } from '../../utilities/storage';

const authToken = getAuthToken();
export const initialState: IAuthStore = {
    isAuthenticated: !!authToken,
    isAuthenticating: false,
    token: authToken
};

function auth(state: IAuthStore = initialState, action: Action): IAuthStore {
    switch (action.type) {
        case 'RECEIVED_LOGOUT':
        case 'FAILED_AUTHENTICATION':
            return {
                isAuthenticated: false,
                isAuthenticating: false,
                token: ''
            };
        case 'REQUEST_AUTHENTICATION':
            return assign({}, state, { isAuthenticating: true });
        case 'RECEIVED_AUTHENTICATION':
        case 'RECEIVED_REGISTRATION':
            return {
                isAuthenticated: true,
                isAuthenticating: false,
                token: action.token
            };
        default:
            return state;
    }
}

export default auth;
