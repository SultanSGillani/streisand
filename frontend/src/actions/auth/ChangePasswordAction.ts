import { replace } from 'react-router-redux';

import Store from '../../store';
import globals from '../../utilities/globals';
import { put } from '../../utilities/Requestor';

import { logout } from './LogoutAction';
import { IUnkownError } from '../../models/base/IError';
import { ThunkAction, IDispatch } from '../ActionTypes';
import ErrorAction, { handleError } from '../ErrorAction';

type ChangePasswordAction =
    { type: 'UPDATING_PASSWORD' } |
    { type: 'UPDATED_PASSWORD' } |
    { type: 'FAILED_PASSWORD_UPDATE' };
export default ChangePasswordAction;
type Action = ChangePasswordAction | ErrorAction;

function updating(): Action {
    return { type: 'UPDATING_PASSWORD' };
}

function updated(): Action {
    return { type: 'UPDATED_PASSWORD' };
}

function failure(): Action {
    return { type: 'FAILED_PASSWORD_UPDATE' };
}

export function changePassword(oldPassword: string, newPassword: string): ThunkAction<Action> {
    return (dispatch: IDispatch<Action>, getState: () => Store.All) => {
        const state = getState();
        dispatch(updating());
        return request(state.sealed.auth.token, oldPassword, newPassword).then((result: { token: string }) => {
            const action = dispatch(updated());
            dispatch(logout());
            dispatch(replace('/login'));
            return action;
        }, (error: IUnkownError) => {
            dispatch(failure());
            return dispatch(handleError(error));
        });
    };
}

function request(token: string, oldPassword: string, newPassword: string): Promise<{ token: string }> {
    const data = { oldPassword, newPassword };
    return put({ token, data, url: `${globals.apiUrl}/change-password/` });
}
