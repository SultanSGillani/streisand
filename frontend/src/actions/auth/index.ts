import AuthAction from './AuthAction';
import LogoutAction from './LogoutAction';
import ChangePasswordAction from './ChangePasswordAction';

type Action = AuthAction | ChangePasswordAction | LogoutAction;
export default Action;