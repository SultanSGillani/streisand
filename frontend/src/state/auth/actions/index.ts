
import LogoutAction from './LogoutAction';
import RegisterAction from './RegisterAction';
import AuthenticateAction from './AuthenticateAction';
import ChangePasswordAction from './ChangePasswordAction';

type Action = AuthenticateAction | ChangePasswordAction | LogoutAction | RegisterAction;
export default Action;