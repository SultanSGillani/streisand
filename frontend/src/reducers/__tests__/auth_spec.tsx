import auth, { initialState } from '../auth';

describe('reducers/auth', () => {
    it ('returns empty state on "RECEIVED_LOGOUT"', () => {
        expect(auth(initialState, {type: 'RECEIVED_LOGOUT'})).toMatchSnapshot();
    });

    it('returns empty state on "FAILED_AUTHENTICATION"', () => {
        expect(auth(initialState, { type: 'FAILED_AUTHENTICATION' })).toMatchSnapshot();
    });

    it('returns state with isAuthenticating: true', () => {
        expect(auth(initialState, { type: 'REQUEST_AUTHENTICATION' })).toMatchSnapshot();
    });

    it('returns state with isAuthenticating: true and token: "12345"', () => {
        expect(auth(initialState, { type: 'RECEIVED_AUTHENTICATION', token: '12345' })).toMatchSnapshot()
    });

    it('returns state with isAuthenticating: true and token: "12345"', () => {
        expect(auth(initialState, { type: 'RECEIVED_REGISTRATION', token: '12345' })).toMatchSnapshot()
    });

    it('returns default state when given un-recognized action', () => {
        expect(auth(initialState, { type: 'UNRECOGNIZED' })).toMatchSnapshot()
    });
});
