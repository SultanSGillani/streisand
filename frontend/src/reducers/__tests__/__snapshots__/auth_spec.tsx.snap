// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/auth returns empty state on "FAILED_AUTHENTICATION" 1`] = `
Object {
  "isAuthenticated": false,
  "isAuthenticating": false,
  "token": "",
}
`;

exports[`reducers/auth returns empty state on "RECEIVED_LOGOUT" 1`] = `
Object {
  "isAuthenticated": false,
  "isAuthenticating": false,
  "token": "",
}
`;

exports[`reducers/auth returns state with isAuthenticating: true 1`] = `
Object {
  "isAuthenticated": false,
  "isAuthenticating": true,
  "token": undefined,
}
`;

exports[`reducers/auth returns state with isAuthenticating: true and token: "12345" 1`] = `
Object {
  "isAuthenticated": true,
  "isAuthenticating": false,
  "token": "12345",
}
`;

exports[`reducers/auth returns state with isAuthenticating: true and token: "12345" 2`] = `
Object {
  "isAuthenticated": true,
  "isAuthenticating": false,
  "token": "12345",
}
`;
